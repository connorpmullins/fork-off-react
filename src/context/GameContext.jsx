import React, { createContext, useContext, useState, useEffect } from "react";
import { db } from "../firebase";
import {
  doc,
  setDoc,
  getDoc,
  updateDoc,
  arrayUnion,
  arrayRemove,
  onSnapshot,
} from "firebase/firestore";

const GameContext = createContext();

export const useGame = () => {
  const context = useContext(GameContext);
  if (!context) {
    throw new Error("useGame must be used within a GameProvider");
  }
  return context;
};

export const GameProvider = ({ children }) => {
  const [gameState, setGameState] = useState({
    roomId: null,
    nickname: null,
    isHost: false,
    players: [],
    gameStarted: false,
  });

  // Subscribe to room updates when in a room
  useEffect(() => {
    if (!gameState.roomId) return;

    const unsubscribe = onSnapshot(
      doc(db, "rooms", gameState.roomId),
      (doc) => {
        if (doc.exists()) {
          const data = doc.data();
          setGameState((prev) => ({
            ...prev,
            players: data.players,
            gameStarted: data.gameStarted,
          }));
        }
      }
    );

    return () => unsubscribe();
  }, [gameState.roomId]);

  const createRoom = async (nickname) => {
    try {
      // Generate a random room ID
      const roomId = Math.random().toString(36).substring(2, 8).toUpperCase();
      const roomRef = doc(db, "rooms", roomId);

      // Create the room document
      await setDoc(roomRef, {
        createdAt: new Date(),
        gameStarted: false,
        players: [
          {
            id: Date.now(),
            nickname,
            isHost: true,
          },
        ],
      });

      setGameState({
        roomId,
        nickname,
        isHost: true,
        players: [
          {
            id: Date.now(),
            nickname,
            isHost: true,
          },
        ],
        gameStarted: false,
      });

      return roomId;
    } catch (error) {
      console.error("Error creating room:", error);
      throw new Error("Failed to create room");
    }
  };

  const joinRoom = async (nickname, roomId) => {
    try {
      const roomRef = doc(db, "rooms", roomId);
      const roomSnap = await getDoc(roomRef);

      if (!roomSnap.exists()) {
        throw new Error("Room not found");
      }

      const roomData = roomSnap.data();
      if (roomData.gameStarted) {
        throw new Error("Game has already started");
      }

      const newPlayer = {
        id: Date.now(),
        nickname,
        isHost: false,
      };

      // Add the new player to the room
      await updateDoc(roomRef, {
        players: arrayUnion(newPlayer),
      });

      setGameState({
        roomId,
        nickname,
        isHost: false,
        players: [...roomData.players, newPlayer],
        gameStarted: roomData.gameStarted,
      });
    } catch (error) {
      console.error("Error joining room:", error);
      throw error;
    }
  };

  const leaveRoom = async () => {
    if (!gameState.roomId) return;

    try {
      const roomRef = doc(db, "rooms", gameState.roomId);
      const currentPlayer = gameState.players.find(
        (p) => p.nickname === gameState.nickname
      );

      if (currentPlayer) {
        await updateDoc(roomRef, {
          players: arrayRemove(currentPlayer),
        });
      }

      // If the leaving player is the host and there are other players,
      // make the next player the host
      if (gameState.isHost) {
        const roomSnap = await getDoc(roomRef);
        const roomData = roomSnap.data();
        const remainingPlayers = roomData.players.filter(
          (p) => p.nickname !== gameState.nickname
        );

        if (remainingPlayers.length > 0) {
          const newPlayers = remainingPlayers.map((p, i) => ({
            ...p,
            isHost: i === 0,
          }));

          await updateDoc(roomRef, { players: newPlayers });
        }
      }

      setGameState({
        roomId: null,
        nickname: null,
        isHost: false,
        players: [],
        gameStarted: false,
      });
    } catch (error) {
      console.error("Error leaving room:", error);
      throw error;
    }
  };

  return (
    <GameContext.Provider
      value={{ gameState, createRoom, joinRoom, leaveRoom }}
    >
      {children}
    </GameContext.Provider>
  );
};
