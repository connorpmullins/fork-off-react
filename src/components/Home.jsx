import React, { useState } from "react";
import { colors, typography, spacing } from "../styles/theme";
import Button from "./common/Button";
import Card from "./common/Card";
import Input from "./common/Input";
import { useGame } from "../context/GameContext";

const Home = () => {
  const { createRoom, joinRoom } = useGame();
  const [nickname, setNickname] = useState("");
  const [roomId, setRoomId] = useState("");
  const [errors, setErrors] = useState({});
  const [isLoading, setIsLoading] = useState(false);

  const containerStyles = {
    minHeight: "100vh",
    backgroundColor: colors.neutral.background,
    padding: spacing[8],
    display: "flex",
    flexDirection: "column",
    alignItems: "center",
  };

  const headerStyles = {
    marginBottom: spacing[8],
    textAlign: "center",
  };

  const titleStyles = {
    color: colors.neutral.gray900,
    fontSize: typography.fontSize["4xl"],
    fontWeight: typography.fontWeight.bold,
    marginBottom: spacing[2],
  };

  const subtitleStyles = {
    color: colors.neutral.gray600,
    fontSize: typography.fontSize.lg,
    maxWidth: "600px",
    marginBottom: spacing[8],
  };

  const formStyles = {
    width: "100%",
    maxWidth: "400px",
  };

  const buttonGroupStyles = {
    display: "grid",
    gridTemplateColumns: "1fr 1fr",
    gap: spacing[4],
    marginTop: spacing[6],
  };

  const validateNickname = () => {
    if (!nickname) {
      return "Nickname is required";
    }
    if (nickname.length < 3) {
      return "Nickname must be at least 3 characters";
    }
    return null;
  };

  const validateRoomId = () => {
    if (!roomId) {
      return "Room ID is required";
    }
    if (roomId.length < 4) {
      return "Room ID must be at least 4 characters";
    }
    return null;
  };

  const handleCreateRoom = async () => {
    const nicknameError = validateNickname();
    if (nicknameError) {
      setErrors((prev) => ({ ...prev, nickname: nicknameError }));
      return;
    }

    setIsLoading(true);
    try {
      await createRoom(nickname);
    } catch (error) {
      setErrors((prev) => ({
        ...prev,
        submit: error.message || "Failed to create room",
      }));
    } finally {
      setIsLoading(false);
    }
  };

  const handleJoinRoom = async () => {
    const nicknameError = validateNickname();
    const roomIdError = validateRoomId();

    if (nicknameError || roomIdError) {
      setErrors({
        nickname: nicknameError,
        roomId: roomIdError,
      });
      return;
    }

    setIsLoading(true);
    try {
      await joinRoom(nickname, roomId);
    } catch (error) {
      let errorMessage = "Failed to join room";
      if (error.message === "Room not found") {
        errorMessage = "Room not found. Please check the room ID.";
      } else if (error.message === "Game has already started") {
        errorMessage = "Cannot join - game has already started.";
      }
      setErrors((prev) => ({ ...prev, submit: errorMessage }));
    } finally {
      setIsLoading(false);
    }
  };

  const handleNicknameChange = (e) => {
    setNickname(e.target.value);
    setErrors((prev) => ({ ...prev, nickname: null }));
  };

  const handleRoomIdChange = (e) => {
    setRoomId(e.target.value);
    setErrors((prev) => ({ ...prev, roomId: null }));
  };

  const getCreateButtonTooltip = () => {
    if (!nickname) return "Please enter a nickname";
    return null;
  };

  const getJoinButtonTooltip = () => {
    if (!nickname) return "Please enter a nickname";
    if (!roomId) return "Please enter a room ID";
    return null;
  };

  return (
    <div style={containerStyles}>
      <header style={headerStyles}>
        <h1 style={titleStyles}>Fork-Off Game</h1>
        <p style={subtitleStyles}>
          Challenge your friends in this exciting multiplayer game where
          strategy meets fun. Create a room or join an existing one to get
          started!
        </p>
      </header>

      <Card>
        <div style={formStyles}>
          <Input
            id="nickname"
            label="Nickname"
            value={nickname}
            onChange={handleNicknameChange}
            error={errors.nickname}
            required
          />

          <Input
            id="room-id"
            label="Room ID (optional for creating)"
            value={roomId}
            onChange={handleRoomIdChange}
            error={errors.roomId}
          />

          {errors.submit && (
            <div style={{ color: colors.error, marginTop: spacing[2] }}>
              {errors.submit}
            </div>
          )}

          <div style={buttonGroupStyles}>
            <Button
              onClick={handleCreateRoom}
              disabled={!nickname || isLoading}
              tooltip={getCreateButtonTooltip()}
              loading={isLoading}
            >
              Create Room
            </Button>

            <Button
              variant="secondary"
              onClick={handleJoinRoom}
              disabled={!nickname || !roomId || isLoading}
              tooltip={getJoinButtonTooltip()}
              loading={isLoading}
            >
              Join Room
            </Button>
          </div>
        </div>
      </Card>
    </div>
  );
};

export default Home;
