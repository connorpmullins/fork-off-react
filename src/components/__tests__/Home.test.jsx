import React from "react";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import Home from "../Home";
import { GameProvider } from "../../context/GameContext";

// Mock functions for GameContext
const mockCreateRoom = jest.fn();
const mockJoinRoom = jest.fn();

// Mock the useGame hook
jest.mock("../../context/GameContext", () => ({
  ...jest.requireActual("../../context/GameContext"),
  useGame: () => ({
    createRoom: mockCreateRoom,
    joinRoom: mockJoinRoom,
    gameState: {
      roomId: null,
      nickname: null,
      isHost: false,
      players: [],
      gameStarted: false,
      phase: "lobby",
    },
  }),
}));

describe("Home", () => {
  beforeEach(() => {
    // Clear mock function calls before each test
    mockCreateRoom.mockClear();
    mockJoinRoom.mockClear();
  });

  const getNicknameInput = () =>
    screen.getByRole("textbox", { name: /nickname/i });
  const getRoomInput = () => screen.getByRole("textbox", { name: /room/i });

  it("renders the home page with all required elements", () => {
    render(<Home />);

    expect(screen.getByText("Fork-Off Game")).toBeInTheDocument();
    expect(getNicknameInput()).toBeInTheDocument();
    expect(getRoomInput()).toBeInTheDocument();
    expect(
      screen.getByRole("button", { name: /create room/i })
    ).toBeInTheDocument();
    expect(
      screen.getByRole("button", { name: /join room/i })
    ).toBeInTheDocument();
  });

  it("matches snapshot of initial state", () => {
    const { container } = render(<Home />);
    expect(container).toMatchSnapshot();
  });

  it("matches snapshot with filled form", () => {
    const { container } = render(<Home />);

    userEvent.type(getNicknameInput(), "testuser");
    userEvent.type(getRoomInput(), "room123");

    expect(container).toMatchSnapshot();
  });

  it("calls createRoom when Create Room button is clicked", async () => {
    render(<Home />);

    userEvent.type(getNicknameInput(), "testuser");
    userEvent.click(screen.getByRole("button", { name: /create room/i }));

    expect(mockCreateRoom).toHaveBeenCalledWith("testuser");
  });

  it("calls joinRoom when Join Room button is clicked", async () => {
    render(<Home />);

    userEvent.type(getNicknameInput(), "testuser");
    userEvent.type(getRoomInput(), "room123");
    userEvent.click(screen.getByRole("button", { name: /join room/i }));

    expect(mockJoinRoom).toHaveBeenCalledWith("testuser", "room123");
  });
});
