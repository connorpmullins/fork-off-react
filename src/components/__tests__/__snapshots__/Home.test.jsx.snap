// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home matches snapshot of initial state 1`] = `
<div>
  <div
    style="min-height: 100vh; background-color: rgb(248, 250, 252); padding: 2rem; display: flex; flex-direction: column; align-items: center;"
  >
    <header
      style="margin-bottom: 2rem; text-align: center;"
    >
      <h1
        style="color: rgb(15, 23, 42); font-size: 2.25rem; font-weight: 700; margin-bottom: 0.5rem;"
      >
        Fork-Off Game
      </h1>
      <p
        style="color: rgb(71, 85, 105); font-size: 1.125rem; max-width: 600px; margin-bottom: 2rem;"
      >
        Challenge your friends in this exciting multiplayer game where strategy meets fun. Create a room or join an existing one to get started!
      </p>
    </header>
    <div
      style="width: 100%; max-width: 900px; display: flex; flex-direction: column; align-items: center;"
    >
      <div
        style="width: 100%; max-width: 400px; margin-bottom: 1.5rem; text-align: center;"
      >
        <div>
          <label
            for="nickname"
            style="display: block; margin-bottom: 0.5rem; color: rgb(51, 65, 85);"
          >
            Your Nickname
            <span
              style="margin-left: 0.25rem;"
            >
              *
            </span>
          </label>
          <input
            id="nickname"
            required=""
            style="width: 100%; padding: 0.75rem; border: 1px solid #cbd5e1; border-radius: 0.375rem; font-size: 1rem; color: rgb(15, 23, 42); outline: none; transition: border-color 0.2s ease;"
            type="text"
            value=""
          />
        </div>
      </div>
      <div
        style="display: grid; grid-template-columns: 1fr 1fr; gap: 1.5rem; width: 100%;"
      >
        <div
          class=""
          role="none"
          style="background-color: rgb(255, 255, 255); border-radius: 0.5rem; padding: 1rem; box-shadow: 0 4px 6px -1px rgb(0 0 0 / 0.1), 0 2px 4px -2px rgb(0 0 0 / 0.1); transition: all 150ms ease-in-out; cursor: default;"
          tabindex="-1"
        >
          <div
            style="height: 100%; display: flex; flex-direction: column;"
          >
            <h2
              style="font-size: 1.25rem; font-weight: 600; margin-bottom: 1rem; color: rgb(15, 23, 42);"
            >
              Create a New Room
            </h2>
            <p
              style="color: rgb(71, 85, 105); margin-bottom: 1rem;"
            >
              Start a new game and invite your friends to join.
            </p>
            <div
              style="margin-top: auto;"
            >
              <div
                data-testid="create-room-button-container"
                style="position: relative; display: inline-block; width: auto;"
              >
                <button
                  class=""
                  disabled=""
                  style="display: inline-flex; align-items: center; justify-content: center; font-family: \\"Inter\\", -apple-system, BlinkMacSystemFont, \\"Segoe UI\\", Roboto, sans-serif; font-weight: 500; border-radius: 0.375rem; transition: all 150ms ease-in-out; cursor: not-allowed; opacity: 0.6; width: 100%; white-space: nowrap; overflow: hidden; text-overflow: ellipsis; outline: none; background-color: rgb(37, 99, 235); color: rgb(255, 255, 255); padding: 0.75rem 1rem; font-size: 1rem; min-width: 120px;"
                  type="button"
                >
                  Create Room
                </button>
              </div>
            </div>
          </div>
        </div>
        <div
          class=""
          role="none"
          style="background-color: rgb(255, 255, 255); border-radius: 0.5rem; padding: 1rem; box-shadow: 0 4px 6px -1px rgb(0 0 0 / 0.1), 0 2px 4px -2px rgb(0 0 0 / 0.1); transition: all 150ms ease-in-out; cursor: default;"
          tabindex="-1"
        >
          <div
            style="height: 100%; display: flex; flex-direction: column;"
          >
            <h2
              style="font-size: 1.25rem; font-weight: 600; margin-bottom: 1rem; color: rgb(15, 23, 42);"
            >
              Join an Existing Room
            </h2>
            <p
              style="color: rgb(71, 85, 105); margin-bottom: 1rem;"
            >
              Enter a room code to join your friends.
            </p>
            <div>
              <label
                for="room-id"
                style="display: block; margin-bottom: 0.5rem; color: rgb(51, 65, 85);"
              >
                Room Code
                <span
                  style="margin-left: 0.25rem;"
                >
                  *
                </span>
              </label>
              <input
                id="room-id"
                required=""
                style="width: 100%; padding: 0.75rem; border: 1px solid #cbd5e1; border-radius: 0.375rem; font-size: 1rem; color: rgb(15, 23, 42); outline: none; transition: border-color 0.2s ease; margin-bottom: 1rem;"
                type="text"
                value=""
              />
            </div>
            <div
              style="margin-top: auto;"
            >
              <div
                data-testid="join-room-button-container"
                style="position: relative; display: inline-block; width: auto;"
              >
                <button
                  class=""
                  disabled=""
                  style="display: inline-flex; align-items: center; justify-content: center; font-family: \\"Inter\\", -apple-system, BlinkMacSystemFont, \\"Segoe UI\\", Roboto, sans-serif; font-weight: 500; border-radius: 0.375rem; transition: all 150ms ease-in-out; cursor: not-allowed; opacity: 0.6; width: 100%; white-space: nowrap; overflow: hidden; text-overflow: ellipsis; outline: none; background-color: rgb(16, 185, 129); color: rgb(255, 255, 255); padding: 0.75rem 1rem; font-size: 1rem; min-width: 120px;"
                  type="button"
                >
                  Join Room
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Home matches snapshot with filled form 1`] = `
<div>
  <div
    style="min-height: 100vh; background-color: rgb(248, 250, 252); padding: 2rem; display: flex; flex-direction: column; align-items: center;"
  >
    <header
      style="margin-bottom: 2rem; text-align: center;"
    >
      <h1
        style="color: rgb(15, 23, 42); font-size: 2.25rem; font-weight: 700; margin-bottom: 0.5rem;"
      >
        Fork-Off Game
      </h1>
      <p
        style="color: rgb(71, 85, 105); font-size: 1.125rem; max-width: 600px; margin-bottom: 2rem;"
      >
        Challenge your friends in this exciting multiplayer game where strategy meets fun. Create a room or join an existing one to get started!
      </p>
    </header>
    <div
      style="width: 100%; max-width: 900px; display: flex; flex-direction: column; align-items: center;"
    >
      <div
        style="width: 100%; max-width: 400px; margin-bottom: 1.5rem; text-align: center;"
      >
        <div>
          <label
            for="nickname"
            style="display: block; margin-bottom: 0.5rem; color: rgb(51, 65, 85);"
          >
            Your Nickname
            <span
              style="margin-left: 0.25rem;"
            >
              *
            </span>
          </label>
          <input
            id="nickname"
            required=""
            style="width: 100%; padding: 0.75rem; border: 1px solid #cbd5e1; border-radius: 0.375rem; font-size: 1rem; color: rgb(15, 23, 42); outline: none; transition: border-color 0.2s ease;"
            type="text"
            value=""
          />
        </div>
      </div>
      <div
        style="display: grid; grid-template-columns: 1fr 1fr; gap: 1.5rem; width: 100%;"
      >
        <div
          class=""
          role="none"
          style="background-color: rgb(255, 255, 255); border-radius: 0.5rem; padding: 1rem; box-shadow: 0 4px 6px -1px rgb(0 0 0 / 0.1), 0 2px 4px -2px rgb(0 0 0 / 0.1); transition: all 150ms ease-in-out; cursor: default;"
          tabindex="-1"
        >
          <div
            style="height: 100%; display: flex; flex-direction: column;"
          >
            <h2
              style="font-size: 1.25rem; font-weight: 600; margin-bottom: 1rem; color: rgb(15, 23, 42);"
            >
              Create a New Room
            </h2>
            <p
              style="color: rgb(71, 85, 105); margin-bottom: 1rem;"
            >
              Start a new game and invite your friends to join.
            </p>
            <div
              style="margin-top: auto;"
            >
              <div
                data-testid="create-room-button-container"
                style="position: relative; display: inline-block; width: auto;"
              >
                <button
                  class=""
                  disabled=""
                  style="display: inline-flex; align-items: center; justify-content: center; font-family: \\"Inter\\", -apple-system, BlinkMacSystemFont, \\"Segoe UI\\", Roboto, sans-serif; font-weight: 500; border-radius: 0.375rem; transition: all 150ms ease-in-out; cursor: not-allowed; opacity: 0.6; width: 100%; white-space: nowrap; overflow: hidden; text-overflow: ellipsis; outline: none; background-color: rgb(37, 99, 235); color: rgb(255, 255, 255); padding: 0.75rem 1rem; font-size: 1rem; min-width: 120px;"
                  type="button"
                >
                  Create Room
                </button>
              </div>
            </div>
          </div>
        </div>
        <div
          class=""
          role="none"
          style="background-color: rgb(255, 255, 255); border-radius: 0.5rem; padding: 1rem; box-shadow: 0 4px 6px -1px rgb(0 0 0 / 0.1), 0 2px 4px -2px rgb(0 0 0 / 0.1); transition: all 150ms ease-in-out; cursor: default;"
          tabindex="-1"
        >
          <div
            style="height: 100%; display: flex; flex-direction: column;"
          >
            <h2
              style="font-size: 1.25rem; font-weight: 600; margin-bottom: 1rem; color: rgb(15, 23, 42);"
            >
              Join an Existing Room
            </h2>
            <p
              style="color: rgb(71, 85, 105); margin-bottom: 1rem;"
            >
              Enter a room code to join your friends.
            </p>
            <div>
              <label
                for="room-id"
                style="display: block; margin-bottom: 0.5rem; color: rgb(51, 65, 85);"
              >
                Room Code
                <span
                  style="margin-left: 0.25rem;"
                >
                  *
                </span>
              </label>
              <input
                id="room-id"
                required=""
                style="width: 100%; padding: 0.75rem; border: 1px solid #cbd5e1; border-radius: 0.375rem; font-size: 1rem; color: rgb(15, 23, 42); outline: none; transition: border-color 0.2s ease; margin-bottom: 1rem;"
                type="text"
                value=""
              />
            </div>
            <div
              style="margin-top: auto;"
            >
              <div
                data-testid="join-room-button-container"
                style="position: relative; display: inline-block; width: auto;"
              >
                <button
                  class=""
                  disabled=""
                  style="display: inline-flex; align-items: center; justify-content: center; font-family: \\"Inter\\", -apple-system, BlinkMacSystemFont, \\"Segoe UI\\", Roboto, sans-serif; font-weight: 500; border-radius: 0.375rem; transition: all 150ms ease-in-out; cursor: not-allowed; opacity: 0.6; width: 100%; white-space: nowrap; overflow: hidden; text-overflow: ellipsis; outline: none; background-color: rgb(16, 185, 129); color: rgb(255, 255, 255); padding: 0.75rem 1rem; font-size: 1rem; min-width: 120px;"
                  type="button"
                >
                  Join Room
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
